caqti-miou/lib/system.ml

```diff
@@ -0,0 +1,237 @@
+
+  let remove_hook hook = Miou.Sequence.remove hook
+
+  let check { orphans; _ } =
+    match Miou.care orphans with
+    | None -> ()
+    | Some None -> ()
+    | Some (Some prm) -> (
+        match Miou.await prm with Ok () -> () | Error _exn -> ())
+end
+
+module Alarm = struct
+  type t = Miou.Condition.t * Miou.Mutex.t
+
+  let schedule ~sw:{ Switch.orphans; _ } ~stdenv:_ t fn =
+    let t_now = Mtime_clock.now () in
+    let mutex = Miou.Mutex.create () and condition = Miou.Condition.create () in
+    let delay =
+      if Mtime.is_later t ~than:t_now then 0.0
+      else Mtime.Span.to_float_ns (Mtime.span t t_now) *. 1e-9
+    in
+    let sleeper = Miou.call_cc @@ fun () -> Miou_unix.sleep delay; `Continue in
+    let canceller =
+      Miou.call_cc @@ fun () ->
+      Miou.Condition.wait condition mutex;
+      `Cancel
+    in
+    let _prm =
+      Miou.call_cc ~orphans @@ fun () ->
+      match Miou.await_first [ sleeper; canceller ] with
+      | Ok `Continue -> fn ()
+      | Ok `Cancel -> ()
+      | Error _exn -> ()
+    in
+    (condition, mutex)
+
+  let unschedule (condition, mutex) =
+    Miou.Mutex.protect mutex @@ fun () -> Miou.Condition.signal condition
+end
+
+let async ~sw:_ fn = Effect.perform (Spawn fn)
+let error_msgf fmt = Fmt.kstr (fun msg -> Error (`Msg msg)) fmt
+
+module Net = struct
+  module Sockaddr = struct
+    type t = Unix.sockaddr
+
+    let unix v = Unix.ADDR_UNIX v
+    let tcp (addr, port) = Unix.ADDR_INET (Ipaddr_unix.to_inet_addr addr, port)
+  end
+
+  let getaddrinfo ~stdenv:() host port =
+    let opts = Unix.[ AI_SOCKTYPE SOCK_STREAM ] in
+    match
+      Unix.getaddrinfo (Domain_name.to_string host) (string_of_int port) opts
+    with
+    | lst -> Ok (List.map (fun ai -> ai.Unix.ai_addr) lst)
+    | exception Not_found -> Ok []
+    | exception Unix.Unix_error (err, f, v) ->
+        error_msgf "%s(%s): %s" f v (Unix.error_message err)
+
+  let convert_io_exception = function
+    | Unix.Unix_error (err, f, v) -> Some (Msg_unix (err, f, v))
+    | _ -> None
+
+  type tcp_flow = Miou_unix.file_descr
+  type tls_flow = |
+
+  module Socket = struct
+    type t = Miou_unix.file_descr
+
+    let output_char fd chr = Miou_unix.write fd (String.make 1 chr) 0 1
+    let output_string fd str = Miou_unix.write fd str 0 (String.length str)
+    let flush _ = ()
+
+    let input_char fd =
+      let buf = Bytes.make 1 '\000' in
+      let len = Miou_unix.read fd buf 0 1 in
+      if len = 0 then raise End_of_file else Bytes.get buf 0
+
+    let really_input fd buf off len =
+      let rec go off = function
+        | 0 -> ()
+        | len ->
+            let len' = Miou_unix.read fd buf off len in
+            go (off + len') (len - len')
+      in
+      go off len
+
+    let close fd = Miou_unix.close fd
+  end
+
+  let connect_tcp ~sw:_ ~stdenv:_ sockaddr =
+    let socket =
+      match sockaddr with
+      | Unix.ADDR_UNIX _ -> invalid_arg "Caqti_miou.Net.connect_tcp"
```

Comment:
In the final version, please use the result type here, with a human readable error.